File: projector/ts-projector/src/projector.ts
import fs from 'node:fs';
import path from 'node:path';
import { Config } from './config';

export type Data = {
  projector: {
    // pwd
    [key: string]: {
      // key          => value
      [key: string]: string
    }
  }
}

const defaultData = {
  projector: {
  }
}

export default class Projector {
  constructor(private config: Config, private data: Data) { }

  getValueAll(): { [key: string]: string } {
    let curr = this.config.pwd;
    let prev = "";

    const paths: string[] = [];
    do {
      prev = curr;
      paths.push(curr);
      curr = path.dirname(curr);
    } while (curr !== prev);
    const result = paths.reverse().reduce((acc, path) => {
      const value = this.data?.projector[path];
      if (value) {
        Object.assign(acc, value);
      }
      return acc;
    }, {});
    return result;
  }

  getValue(key: string): string | undefined {
    let curr = this.config.pwd;
    let prev = "";

    let out: string = "";

    do {
      const value = this.data?.projector[curr]?.[key];
      if (value) {
        out = value;
        break;
      }
      prev = curr;
      curr = path.dirname(curr);
    } while (curr !== prev);

    return out;
  }

  setValue(key: string, value: string) {
    let dir = this.data?.projector[this.config.pwd];
    if (!dir) {
      dir = {};
      this.data.projector[this.config.pwd] = dir;
    }
    dir[key] = value;
  }

  removeValue(key: string) {
    let dir = this.data?.projector[this.config.pwd];
    if (dir) {
      delete dir[key];
    }
  }

  save() {
    const configPath = path.dirname(this.config.config);
    if (!fs.existsSync(configPath)) {

      fs.mkdirSync(configPath, { recursive: true });
    }
    fs.writeFileSync(this.config.config, JSON.stringify(this.data));
  }

  static fromConfig(config: Config): Projector {
    let data: Data = { projector: {} };
    if (fs.existsSync(config.config)) {
      try {
        data = JSON.parse(fs.readFileSync(config.config, 'utf-8').toString());
      } catch (error) {
        data = defaultData;
        return new Projector(config, data);
      }
    }
    return new Projector(config, data);
  }
}

File: fm-cli/go-projector/pkg/projector/projector.go
package projector

import (
	"encoding/json"
	"os"
	"path"
)

type Data struct {
	Projector map[string]map[string]string `json:"projector"`
}

type Projector struct {
	config *Config
	data   *Data
}

func CreateProjector(config *Config, data *Data) *Projector {
	return &Projector{
		config: config,
		data:   data,
	}
}

func (p *Projector) GetValue(key string) (string, bool) {
	curr := p.config.Pwd
	prev := ""
	result := ""
	for curr != prev {
		if value, ok := p.data.Projector[curr][key]; ok {
			result = value
			break
		}
		prev = curr
		curr = path.Dir(curr)
	}
	return result, result != ""
}

func (p *Projector) GetAllValue() map[string]string {
	curr := p.config.Pwd
	prev := ""
	results := map[string]string{}

	for curr != prev {
		if dir, ok := p.data.Projector[curr]; ok {
			for key, value := range dir {
				results[key] = value
			}
		}
		prev = curr
		curr = path.Dir(curr)
	}

	return results
}

func (p *Projector) SetValue(key, value string) {
	if _, ok := p.data.Projector[p.config.Pwd]; !ok {
		p.data.Projector[p.config.Pwd] = map[string]string{key: value}
	}
	p.data.Projector[p.config.Pwd][key] = value
}

func (p *Projector) RemoveValue(key string) {
	if _, ok := p.data.Projector[p.config.Pwd]; ok {
		delete(p.data.Projector[p.config.Pwd], key)
	}
}

func (p *Projector) Save() error {
	dir := path.Dir(p.config.Config)
	if _, err := os.Stat(dir); os.IsNotExist(err) {
		err = os.MkdirAll(dir, 0755)
		if err != nil {
			return err
		}
	}

	contents, err := json.Marshal(p.data)
	if err != nil {
		return err
	}
	return os.WriteFile(p.config.Config, contents, 0755)
}

func defaultProjector(config *Config) *Projector {
	return &Projector{
		config: config,
		data: &Data{
			Projector: map[string]map[string]string{},
		},
	}
}

func NewProject(config *Config) *Projector {
	if _, err := os.Stat(config.Config); err == nil {
		contents, err := os.ReadFile(config.Config)
		if err != nil {
			return defaultProjector(config)
		}

		var data Data

		err = json.Unmarshal(contents, &data)
		if err != nil {
			return defaultProjector(config)
		}

		return &Projector{
			data:   &data,
			config: config,
		}
	}
	return defaultProjector(config)
}
